source/slang/slang-check-modifier.cpp

```diff
@@ -1228,6 +1228,37 @@ namespace Slang
             }
         }
 
+        MemoryQualifierCollectionModifier::Flags::MemoryQualifiersBit memoryQualifierBit = 
+            MemoryQualifierCollectionModifier::Flags::kNone;
+        if(as<GloballyCoherentModifier>(m))
+            memoryQualifierBit = MemoryQualifierCollectionModifier::Flags::kCoherent;
+        else if(as<GLSLReadOnlyModifier>(m))
+            memoryQualifierBit = MemoryQualifierCollectionModifier::Flags::kReadOnly;
+        else if(as<GLSLWriteOnlyModifier>(m))
+            memoryQualifierBit = MemoryQualifierCollectionModifier::Flags::kWriteOnly;
+        else if(as<GLSLVolatileModifier>(m))
+            memoryQualifierBit = MemoryQualifierCollectionModifier::Flags::kVolatile;
+        else if(as<GLSLRestrictModifier>(m))
+            memoryQualifierBit = MemoryQualifierCollectionModifier::Flags::kRestrict;
+        if(memoryQualifierBit != MemoryQualifierCollectionModifier::Flags::kNone)
+        {
+            bool newModifier = false;
+            MemoryQualifierCollectionModifier* memoryQualifiers = syntaxNode->findModifier<MemoryQualifierCollectionModifier>();
+            if(!memoryQualifiers)
+            {
+                newModifier = true;
+                memoryQualifiers = getASTBuilder()->create<MemoryQualifierCollectionModifier>();
+            }
+            memoryQualifiers->addQualifier(m,
+                memoryQualifierBit);
+            if (newModifier)
+            {
+                m->next = memoryQualifiers;
```

Comment:
`next` will be overwritten by `checkModifiers` so this assignment will have no effect. We should just delete this line.

Reply:
I need to return multiple modifiers (current+next) to keep compatibility with other code. 

I will add to checkMemoryQualifiers to solve this.

Reply:
It is safe to just return `nullptr` here, in that case the original modifier won't appear in the checked modifier list.

Reply:
checked modifiers replace the modifiers list in use. Since we need the memory qualifier and container, both need to be added

Reply:
edit: I don't need to handle 2 modifiers, that was a misjudgment of the code, I have another way I can do the implementation

Reply:
I think you can just add the original modifier to the list inside the new modifier, and return the new modifier.

Reply:
The original modifier does not need to exist in the checked modfiier list, as long as our ir-lowering logic is no longer looking for those original modifiers. We keep them in the list of the new modifier just in case we need their source loc to report diagnostics or in case language server needs them.